core/lib/Thelia/Controller/Admin/ProductController.php

```diff
@@ -1804,4 +1816,365 @@ protected function arrayHasEntries(array $data, array $entries)
+            ->findOneByProductId($cpId);
+        $cpDefaultPSE
+            ->setQuantity($opPSE->getQuantity())
+            ->setPromo($opPSE->getPromo())
+            ->setNewness($opPSE->getNewness())
+            ->setEanCode($opPSE->getEanCode())
+            ->save();
+
+        // Set default promo price
+        $cpDefaultPrice = ProductPriceQuery::create()
+            ->findOneByProductSaleElementsId($cpDefaultPSE->getId());
+        $cpDefaultPrice
+            ->setPromoPrice($opPromoPrice)
+            ->save();
+
+        return ['cpId' => $cpId, 'cpDefaultPSEId' => $cpDefaultPSE->getId()];
+    }
+
+    public function setCloneAttributeCombination($opPSEId, $cpPSEId)
+    {
+        // Get original product attribute for current PSE
+        $opAttrCombi = AttributeCombinationQuery::create()
+            ->findOneByProductSaleElementsId($opPSEId);
+
+        $cpAttrCombi = new AttributeCombination();
+        $cpAttrCombi
+            ->setAttributeId($opAttrCombi->getAttributeId())
+            ->setAttributeAvId($opAttrCombi->getAttributeAvId())
+            ->setProductSaleElementsId($cpPSEId)
+            ->save();
+    }
+
+    public function setCloneFeatureCombination($opId, $cpId)
+    {
+        // Get original product features
+        $opFeatures = FeatureProductQuery::create()
+            ->findByProductId($opId);
+
+        // Set clone product features
+        foreach ($opFeatures as $opFeature) {
+            $cpFeatureProduct = new FeatureProduct();
+            $cpFeatureProduct
+                ->setProductId($cpId)
+                ->setFeatureId($opFeature->getFeatureId())
+                ->setFeatureAvId($opFeature->getFeatureAvId())
+                ->setFreeTextValue($opFeature->getFreeTextValue())
+                ->setPosition($opFeature->getPosition())
+                ->save();
+        }
+    }
+
+    public function setProductAssociatedContent($opId, $cpId)
+    {
+        // Get original product associated contents
+        $opAssocConts = ProductAssociatedContentQuery::create()
+            ->findByProductId($opId);
+
+        // Set clone product associated contents
+        foreach ($opAssocConts as $opAssocCont) {
+            $cpProdAssocCont = new ProductAssociatedContent();
+            $cpProdAssocCont
+                ->setProductId($cpId)
+                ->setContentId($opAssocCont->getContentId())
+                ->setPosition($opAssocCont->getPosition())
+                ->save();
+        }
+    }
+
+    public function setCloneI18n($lang, $opId, $cpId)
+    {
+        // Get original product i18n
+        $opI18ns = ProductI18nQuery::create()
+            ->findById($opId);
+
+        /* For each original I18n
+         * - If it's the same language as the one at the product creation, update information
+         * - If not, create a new ProductI18n for the language
+         */
+        foreach ($opI18ns as $opI18n) {
+            if ($opI18n->getLocale() == $lang) {
```

Comment:
Have a look at Propel's copy() method, it may saves you from writing tedious code.